pkg/controller/utils/utils.go

```diff
@@ -471,3 +423,59 @@ func StrToElasticLicenseType(license string, logger logr.Logger) render.Elastics
 	logger.V(3).Info("Elasticsearch license %s is unexpected", license)
 	return render.ElasticsearchLicenseTypeUnknown
 }
+
+// WaitToAddResourceWatch will check if projectcalico.org APIs are available and if so, it will add a watch for resource
+// The completion of this operation will be signaled on a ready channel
+func WaitToAddResourceWatch(controller controller.Controller, client kubernetes.Interface, log logr.Logger, flag *ReadyFlag, resourceKind string) {
```

Comment:
I think that if we're going to make this more generic then we could just go all the way and have the ability to run an arbitrary function based off of an arbitrary condition that can be set by the caller. For instance, a crude implementation would be:

```
func WaitToExecute(f1 func(), f2 func() bool, done chan struct{}) {
	maxDuration := 30 * time.Second
	duration := 1 * time.Second
	ticker := time.NewTicker(duration)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			duration = duration * 2
			if duration >= maxDuration {
				duration = maxDuration
			}
			ticker.Reset(duration)

			if f2() {
				f1()
				close(done)
			}
		}
	}
}
```

Then you could just have an exported `IsResourceReady`, and in your controller you would just run:
```
done := make(chan struct{})
	go utils.WaitToExecute(func() {
		controller.Watch(&source.Kind{Type: &v3.DeepPacketInspection{TypeMeta: metav1.TypeMeta{Kind: v3.KindDeepPacketInspection}}},
			&handler.EnqueueRequestForObject{})
	}, func() bool {
		return utils.IsResourceReady(client, v3.KindDeepPacketInspection)
	}, done)
```

It can be used in any situation where we need to delay watching based off of any arbitrary signal, it's very readable what's happening in the controller.

Reply:
I think I'll leave it the way it is for now, probably better to do it a separate PR.

Reply:
Sounds good!